File: src/main/java/stack/RemoveAllAdjacentDuplicatesinStringII.java
package stack;
import java.util.Stack;
public class RemoveAllAdjacentDuplicatesinStringII {
	public String removeDuplicates(String s, int k) {
        if ( s.length() < 2 ) return s;
        char[] ca = s.toCharArray();
        Stack<Pair> stack = new Stack<>();
        stack.push(new Pair(ca[0],1));
        for (int i = 1; i < ca.length; i++) {
            if ( ! stack.isEmpty() ) {
                char c = stack.peek().c;
                int len = stack.peek().len;
                //System.out.println(c + " " + len);
                if ( c == ca[i] ) { // A repeating character
                    len = len + 1;
                    if ( len == k ) {
                        for(int l = 1; l < k; l++) {
                            stack.pop();
                        }
                    }
                    else {
                        stack.push(new Pair(c,len));
                    }
                }
                else {             //A new character
                    stack.push(new Pair(ca[i],1));
                }
            } 
            else {
                stack.push(new Pair(ca[i],1));
            }
            
        }
        StringBuilder sb = new StringBuilder();
        while ( ! stack.isEmpty() ) {
            sb.insert(0,stack.pop().c);
        }
        return sb.toString();
    }
    
    class Pair {
        public char c;
        public int len;
        public Pair(char _c, int _len) {
            c = _c;
            len = _len;
        }
    }
}

File: Remove All Adjacent Duplicates in String II.py
class Solution:
    def removeDuplicates(self, s: str, k: int) -> str:
        stack = []
        for c in s:
            if stack and stack[-1][0] == c:
                stack[-1][1] += 1
            else:
                stack.append( [c, 1] )
            if stack[-1][1] == k:
                stack.pop()
        res = []
        for i in range(len(stack)):
            for j in range(stack[i][1]):
                res.append(stack[i][0])
        return ''.join(res)
